### PR TITLE
RED-26361 CRDB Recovery with AOF

### PR DESCRIPTION
**Pull request by bmansheim**

### CODE REVIEW

**Comment by bmansheim** [comment]
@ushachar I rewrote the process for cluster recovery. Who can review it (high level) to make sure I didn't change any technical details?
Once it is reviewed I will add pieces that will relate to recovery of CRDB.

**Comment by bmansheim** [comment]
@ushachar @miki@redislabs.com @Paz-Yanover In an attempt to make cluster recovery more understandable, I rewrote the information and split it into two pages: cluster recovery and database recovery.

Please review the content of the pages and add your comments to:

- https://github.com/RedisLabs/redislabs-docs/pull/384/files#diff-67d875aebd972c9c0cbd45907c57c703 (https://docs.redislabs.com/staging/RED-26361-crdb-recovery/rs/administering/troubleshooting/cluster-recovery/)
- https://github.com/RedisLabs/redislabs-docs/pull/384/files#diff-e562f397073af704c373b24aafc1cf38 (https://docs.redislabs.com/staging/RED-26361-crdb-recovery/rs/administering/troubleshooting/database-recovery/)

**Comment by ushachar** [comment]
@otnielvh  - Can you take a look at the above?

**Comment by bmansheim** [comment]
@otnielvh I already added the content for CRDB recovery but it is only lines 80-84 in the database-recovery article.

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 1]
```diff
@@ -1,215 +1,146 @@
...
-created by the cluster. These files are stored in the persistence
-storage location (for additional details, refer to [Persistent and
-ephemeral
-storage]({{< relref "/rs/administering/designing-production/persistent-ephemeral-storage.md" >}}).
-
-Prerequisites:
-
-- You will need the persistent storage drives of the old cluster.
+When a cluster fails,
+you must recover the database with its previous cluster configuration, databases and data.
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: There are two distinguish processes: recovering the cluster configuration (i.e. the node numbering, the BDB configuration and placement etc.), and recovering the user data (the information stored inside the BDBs). I would try to make this clearer in the documentation.

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 2]
```diff
@@ -1,215 +1,146 @@
...
+To recover the cluster to the state that it was in before the failure,
+you must create a new cluster that has the same node configuration as the original cluster.
+
+Cluster recovery requires the cluster configuration backup files and database persistence files
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: What's the difference between "backup files" and "persistence files"? I know what you mean, but I think the user won't understand

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 3]
```diff
@@ -1,215 +1,146 @@
...
+that are created by the cluster.
+These files are stored in the [persistence storage location]
+({{< relref "/rs/administering/designing-production/persistent-ephemeral-storage.md" >}}).
+
+The cluster recovery process includes:
+
+1. Install RS on the nodes of the new cluster.
+1. Mount to the new nodes the storage that holds the cluster configuration and persistent storage of the original cluster.
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: The sentence is correct but a bit tung twisting. Better wording would be something like: "Mount the original cluster files (both cluster configuration and BDB backups) to the new cluster's nodes"

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 4]
```diff
@@ -1,215 +1,146 @@
...
-configuration backup files and database persistence files that are
-created by the cluster. These files are stored in the persistence
-storage location (for additional details, refer to [Persistent and
-ephemeral
-storage]({{< relref "/rs/administering/designing-production/persistent-ephemeral-storage.md" >}}).
-
-Prerequisites:
-
-- You will need the persistent storage drives of the old cluster.
+When a cluster fails,
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: What does "cluster failure" mean? I know this is a pandora box, but we might want to address it

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 5]
```diff
@@ -1,215 +1,146 @@
...
-fail. Make sure to kill all Redis processes before attempting to run the
-recovery process.
+    database persistence files.
+
+    {{% note %}}
+Make sure that the user redislabs has permissions to access the storage location
+of the configuration and persistence files on each of the nodes.
+    {{% /note %}}
+
+    If you use local persistent storage, place all the recovery files on each of the cluster nodes.
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: "place all the recovery files" - only the relevant recovery files are required. I.e. the CCS for the first node, and the BDB backup files for the rest.

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 6]
```diff
@@ -1,215 +1,146 @@
...
+
+1. To recover the cluster with the configuration from the original cluster,
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: "To begin the recovering process ... ". At the end of this process the cluster is not configured yet, this step only initiates the process

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 7]
```diff
@@ -1,215 +1,146 @@
...
+    from the [rladmin]({{< relref "/rs/references/cli-reference/rladmin.md" >}}) command-line interface (CLI):
+
+    ```src
+    cluster recover filename [ <persistent_path> | <ephemeral_path> ]<filename> node_uid <node_uid> rack_id <rack_id>
+    ```
+
+    {{% expand "Command syntax" %}}
+`<filename>` - The full path of the old cluster configuration file in the persistent storage.
+The cluster configuration file is /css/ccs-redis.rdb.
+The file exists on the persistent storage drive of all nodes and all copies are identical.
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: You mean exists on the old cluster...

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 8]
```diff
@@ -1,215 +1,146 @@
...
+will get the same rack ID as the original node.
+    {{% /expand %}}
+
+    For example:
+
+    ```src
+    rladmin cluster recover filename /tmp/persist/ccs/ccs-redis.rdb node_uid 1 rack_id 5
+    ```
+
+    When the recovery succeeds,
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: Recovery process was just initiated. When this phase is done, the recovery isn't completed yet.

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 9]
```diff
@@ -1,215 +1,146 @@
...
+    this node is configured as the node with ID 1 from the old cluster.
+
+1. To join the remaining servers to the recovered cluster, from the rladmin CLI of each new node run:
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/cluster-recovery.md]: "new cluster" versus "recovered cluster"

File: content/rs/administering/troubleshooting/database-recovery.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+After you prepare the cluster that hosts the database,
+you can run the recovery process from the [rladmin]({{< relref "/rs/references/cli-reference/rladmin.md" >}})
+command-line interface (CLI).
+
+{{% note %}}
+Before you run the recovery process on the machines from the original cluster with Redis processes running on the machine,
+make sure to kill all Redis processes.
+{{% /note %}}
+
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/database-recovery.md]: I don't think this is needed for database recovery, only for cluster recovery.

File: content/rs/administering/troubleshooting/database-recovery.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    {{% note %}}
+If the persistence files of the databases from the old cluster are not stored in the persistent storage location of the new node,
+you must first map the recovery path of each node to the location of the old persistence files.
+To do this, run the `node <id> recovery_path set` command in rladmin.
+The persistence files for each database are located in the persistent storage path of the nodes from the old cluster, under the /redis directory.
+    {{% /note %}}  
+
+1. To verify that the recovered databases are now active, run: `rladmin status`
+
+After the databases are recovered, make sure that your redis clients can successfully connect to the cluster.
```
Comment (otnielvh) [inline content/rs/administering/troubleshooting/database-recovery.md]: "... can successfully connect to the cluster" should be "... can successfully connect to the database"

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 3 above)
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: How's this?

> Mount the persistent storage from the original cluster to the new nodes of the new cluster.

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 12]
```diff
@@ -1,215 +1,146 @@
...
+## Prerequisites
+
+- Before you run the recovery process on the machines from the original cluster with Redis processes running on the machine,
+make sure to kill all Redis processes.
+
+- We recommend that you allocate new persistent storage drives for the new cluster nodes.
+If you decide to use the persistent storage drives of the old cluster nodes,
+make sure that you backup all files on the old persistent storage drives to another location.
+
+## Recovering the Cluster
```
[the remark this answers is not in the document]
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: This is standard writing practice for me but I do see your point.

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 13]
```diff
@@ -1,215 +1,146 @@
...
+1. Recover the cluster configuration on the first node in the new cluster.
+1. Join the remaining nodes to the new cluster.
+1. [Recover the databases]({{< relref "/rs/administering/troubleshooting/database-recovery.md" >}}).
+
+## Prerequisites
+
+- Before you run the recovery process on the machines from the original cluster with Redis processes running on the machine,
+make sure to kill all Redis processes.
+
+- We recommend that you allocate new persistent storage drives for the new cluster nodes.
```
[the remark this answers is not in the document]
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: I tried write it more clearly in my latest commit. You can look at it and tell me if you think it's better.

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 14]
```diff
@@ -1,215 +1,146 @@
...
+1. Install RS on the nodes of the new cluster.
+1. Mount to the new nodes the storage that holds the cluster configuration and persistent storage of the original cluster.
+1. Recover the cluster configuration on the first node in the new cluster.
+1. Join the remaining nodes to the new cluster.
+1. [Recover the databases]({{< relref "/rs/administering/troubleshooting/database-recovery.md" >}}).
+
+## Prerequisites
+
+- Before you run the recovery process on the machines from the original cluster with Redis processes running on the machine,
+make sure to kill all Redis processes.
```
[the remark this answers is not in the document]
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: I tried write it more clearly in my latest commit. You can look at it and tell me if you think it's better.

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 15]
```diff
@@ -1,215 +1,146 @@
...
+
+- We recommend that you allocate new persistent storage drives for the new cluster nodes.
+If you decide to use the persistent storage drives of the old cluster nodes,
+make sure that you backup all files on the old persistent storage drives to another location.
+
+## Recovering the Cluster
+
+1. Install [RS]({{< relref "/rs/installing-upgrading/downloading-installing.md" >}}) on the servers
+    for the new cluster nodes. Do not configure the cluster, also called cluster setup.
```
[the remark this answers is not in the document]
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: How's this?

> Do not configure the cluster nodes (`rladmin cluster create` in the CLI or **Setup** in the Web UI).

File: content/rs/administering/troubleshooting/cluster-recovery.md [hunk 16]
```diff
@@ -1,215 +1,146 @@
...
+
+    The new servers must have the same basic hardware and software configuration as the original servers, including:
+
+    - The same number of nodes
+    - At least the same memory available on each node
```
[the remark this answers is not in the document]
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: How's this?

> At least the same amount of memory

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 6 above)
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: How's this?

> To recover the cluster configuration from the original cluster to the first node in the new cluster,

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 7 above)
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: How's this?

> The file is located on the persistent storage drive for each node in the original cluster and all copies are identical.

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 8 above)
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: How's this?

> When the recovery command succeeds,

File: content/rs/administering/troubleshooting/database-recovery.md (hunk 10 above)
Comment (bmansheim) [reply content/rs/administering/troubleshooting/database-recovery.md]: Do you want to confirm it or should I remove it?

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 5 above)
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: Are the "database persistence files" just the backup files that you create with export or the scheduled backup process? What about AOF or snapshots?

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 2 above)
Comment (otnielvh) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: Both the cluster configuration and the user data are both stored in "persistent backup files", I think the distinction should be between configuration data and user data. Also, we should be clear on the terminology, and I use R&D terminology. You'll want to use customer facing terminology, and that you will get better from the marketing guys

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 3 above)
Comment (otnielvh) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: "..new nodes of the new cluster" doubles the new. I think you can do without the first new.

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 4 above)
Comment (otnielvh) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: I will let @ushachar answer on this

Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: @ushachar I got the quorum loss point from Miki.

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (otnielvh) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: Oh, you mean "Do not create the cluster", you are going to configure the cluster nodes though.

File: content/rs/administering/troubleshooting/database-recovery.md (hunk 10 above)
Comment (otnielvh) [reply content/rs/administering/troubleshooting/database-recovery.md]: Not sure how to verify it. You can maybe ask someone from support.

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: Right

File: content/rs/administering/troubleshooting/cluster-recovery.md (hunk 4 above)
Comment (bmansheim) [reply content/rs/administering/troubleshooting/cluster-recovery.md]: @ushachar Last chance before I publish.

**Pull request merged**